Index: ofed_kernel-2.6.9_U8/drivers/infiniband/ulp/sdp/sdp_zcopy.c
===================================================================
--- ofed_kernel-2.6.9_U8.orig/drivers/infiniband/ulp/sdp/sdp_zcopy.c	2010-10-27 15:21:18.000000000 +0200
+++ ofed_kernel-2.6.9_U8/drivers/infiniband/ulp/sdp/sdp_zcopy.c	2010-10-27 15:29:18.000000000 +0200
@@ -405,7 +405,7 @@
 	if (capable(CAP_IPC_LOCK))
 		return ULONG_MAX;
 
-	lock_limit = current->signal->rlim[RLIMIT_MEMLOCK].rlim_cur;
+	lock_limit = current->rlim[RLIMIT_MEMLOCK].rlim_cur;
 	avail = lock_limit - (current->mm->locked_vm << PAGE_SHIFT);
 
 	return avail < offset ? 0 : avail - offset;
@@ -449,8 +449,8 @@
 		rc = -EAGAIN;
 		sdp_dbg_data(sk, "Error doing umem_get 0x%zx bytes: %d\n", len, rc);
 		sdp_dbg_data(sk, "RLIMIT_MEMLOCK: 0x%lx[cur] 0x%lx[max] CAP_IPC_LOCK: %d\n",
-				current->signal->rlim[RLIMIT_MEMLOCK].rlim_cur,
-				current->signal->rlim[RLIMIT_MEMLOCK].rlim_max,
+				current->rlim[RLIMIT_MEMLOCK].rlim_cur,
+				current->rlim[RLIMIT_MEMLOCK].rlim_max,
 				capable(CAP_IPC_LOCK));
 		goto err_umem_get;
 	}
