--- dhcp.spec.ib	2009-07-02 10:27:27.000000000 +0300
+++ dhcp.spec	2010-09-20 12:48:15.559809000 +0200
@@ -52,6 +52,9 @@ Patch17: %{name}-3.0.5-RES_OPTIONS.patch
 Patch18: %{name}-3.0.5-emergency-relay-agent-options-hologram.patch
 Patch19: %{name}-3.0.5-honor-peer-vars.patch
 Patch20: %{name}-3.0.5-IFNAMSIZ.patch
+Patch21: %{name}-3.0.5-lpf-ib.patch
+Patch22: %{name}-3.0.5-improved-xid.patch
+Patch23: %{name}-3.0.5-gpxe-cid.patch
 
 BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
 Requires(post): chkconfig, coreutils
@@ -198,6 +200,12 @@ esac
 # dhcpd (RHBZ #441524)
 %patch20 -p1 -b .ifnamsiz
 
+%patch21 -p1 -b .ib
+
+%patch22 -p1 -b .xid
+
+%patch23 -p1 -b .gpxe
+
 %build
 %{__cp} %SOURCE1 .
 %{__cat} << EOF > site.conf
