diff --git a/drivers/infiniband/hw/qib/qib.h b/drivers/infiniband/hw/qib/qib.h
index 305f9b6..3707d79 100644
--- a/drivers/infiniband/hw/qib/qib.h
+++ b/drivers/infiniband/hw/qib/qib.h
@@ -1405,8 +1405,7 @@ extern struct mutex qib_mutex;
 #define qib_early_err(dev, fmt, ...) \
 	do { \
 		qib_log(__QIB_INFO, "EarlyErr: " fmt, ##__VA_ARGS__); \
-		dev_info(dev, KERN_ERR QIB_DRV_NAME ": " fmt, \
-			##__VA_ARGS__); \
+		dev_err(dev, fmt, ##__VA_ARGS__); \
 	} while (0)
 
 #define qib_dev_err(dd, fmt, ...) \
